### ISSUE TITLE
window.zoomLevel causing blurry canvas

### ISSUE THREAD

**Issue by ZZYSonny** [microsoft/vscode#164014]
Does this issue occur when all extensions are disabled?: Yes

- VS Code Version: 1.72.2
- OS Version: Arch Linux, GNOME 42.4

Steps to Reproduce: See https://github.com/James-Yu/LaTeX-Workshop/issues/3509
1. Set "window.zoomLevel" to 2 in VSCode settings.
2. Start a page with canvas. (For example vscode-pdf uses pdf-js, which uses canvas to render a page)
3. Toggle developer tools and run `window.devicePixelRatio` in the console.

Screenshot:
You can see the second 123 is clearer than the first one.
`ZoomLevel=2`
![image](https://user-images.githubusercontent.com/28442838/196675511-10e6d2a7-296c-4d91-a6a4-a7c07b0839c8.png)
`ZoomLevel=log(1.5,1.2)`
![image](https://user-images.githubusercontent.com/28442838/196676061-90ca3ab9-8872-4fe4-8128-faf607d19edc.png)

Problem this is causing:
From what I observed, VSCode is likely using a exponential scaling `window.devicePixelRatio = power(1.2, window.zoomLevel)`. While this makes zooming intuitive, it is likely to cause roundup issue. With zoomLevel=2, devicePixelRatio is 1.44 (up to some floating rounding issue). When one set canvas size using the common `width * window.devicePixelRatio` way, it is more likely to miss a few pixels and cause Chrome to downscale or upscale the canvas.

Possible workaround:
Set `window.zoomLevel` to `log(1.5,1.2)`. So `window.devicePixelRatio` will be exactly 1.5.

Possible solution:
Chrome only provides fixed page zoom ratio. Maybe we can round the exponential zoom level to one of the fixed zoom level. (1.44 rounds to 1.5) 
![image](https://user-images.githubusercontent.com/28442838/196674541-9aa423d3-770b-4141-9c73-c7faabed140b.png)

**Comment by bpasero** [comment]
VS Code zoom goes directly to Electron, so any change needed should probably happen there for all apps to benefit from.

**Comment by ZZYSonny** [comment]
"window.zoomLevel" in VSCode setting seems to correspond to `webFrame.setZoomLevel(level)` in electron. (which gives you the 1.2^x scaling and causes floating round issue)

Electron also seems to support `webFrame.setZoomFactor`, which gives you (x/100) as the scaling. I think with this it is easier to set a zoomFactor 1,1.5,2 which causes less rounding issues.

**Comment by bpasero** [comment]
Right, thanks for pointing out. I am a bit afraid of messing with how we do zooming to be honest.
